### ISSUE TITLE
CRUCIAL BUG: Autosave session stopped working

### ISSUE THREAD

**Issue by mockinbirdy** [sienori/Tab-Session-Manager#1384]
The autosave session function stopped working, it no longer overwrites old sessions and instead just stops saving new sessions altogether defeating the sole purpose of the extension. 

Again, it's been reported #1367 in July and it stopped working in early June. I bet 90% of users don't even know about this BUG and what a big surprise awaits them once they open this extension to look for one of their last sessions.  

According to people's complains, it affects at least Edge, Chrome and Firefox in my case. So I guess it affects every other browser as well. I needs to be fixed asap.

**Comment by freemedom** [comment]
i think this is because of the upgrade for v3.

**Comment by nycal99** [comment]
I DON'T UNDERSTAND WHY YOU DON'T WANT TO FIX THIS. Firefox for me too.

I JUST DISCOVERED THIS SUMMER AND AM VERY UNHAPPY. Yes. indeed it was a "big surprise."

Is there another Add-on for Foxfire that does something similar? I am ready to switch after this experience.

**Comment by lrq3000** [comment]
Same here, and the extension cannot show nor save any session whatsoever. I just found out my last autoupdate (saved in different files with random names so there is no overwriting issue on my side, and all in one json file per save) happened 2 weeks ago...

But I did not realize until today, because all tabs that were unloaded fr m memory (using The Marvellous Suspender) are now just empty and cannot be recovered, there is no URL, no content whatsoever.

So I just lost my last 2 weeks of content to browse or process. Fortunately I heavily worked on other devices where I had separate softwares to keep track of my most important material, but still it's problematic, especially since the extension does not work at all anymore.

I don't know exactly why the autoupdate stopped working, but the empty pages seem to have started yesterday with an autoupdate of Chrome to 130.0.6723.58 , which greets you with the new chrome tabs features (you can see tabs groups below the search bar on the homepage). I think this conflicted with Tabius, an extension I have that already autogroups tabs together, which is what Chrome intends to do now.

It seems Google did not expect/care for those who already were using tabs groups extensively since a while. When I look in Chrome's tabs groups list, I see that the empty pages are titles "Unsavable tab". Apparently, we are not alone, and this corruption of tabs happens even to [those who are using vanilla chrome](https://www.reddit.com/r/chrome/comments/1b5dqnr/do_not_save_group_tabs_they_might_be_replaced/), it seems Google just pushed a very unstable feature.

**Comment by ToyKeeper** [comment]
New user here.  Just switched to Firefox, added TST, added Tab Session Manager, and I have it configured mostly default, like to auto-save every 15 minutes...  and I soon found myself wondering:

Is it supposed to save *to regular files on the filesystem?*

Because it's not doing that.

And if it doesn't do that, what is the point?

The only function I need is a periodic automatic export of my session, to a local file, in a standard text format like json.

After using it for a full day, and closing and reopening the browser a few times to test things, and attempting to use TSM's manual save function a few times... it did actually dump a few session files under `~/Downloads/TabSessionManager-Backup/` ... but only once, and the sessions it dumped were already 9 to 16 hours out of date at the time they were written to disk.  And I haven't been able to get it to save to files again since then.  Not even with manual saves.  And it didn't even use the directory name I gave it, or the date format I told it to use.

So it did *something* right ... once.  And looking at the files it saved, it appears to have the correct(ish) data inside for the times stamped in the filenames.  So now I'm wondering if I need to fork my own modded version or something, which would send the data to a separate web service on localhost, and write a minimal web service to just accept data and write it to disk.  Maybe that would get around whatever was preventing it from working.

(ish: I found broken tree structures in 3 of the session files it exported, but I think those were broken in Firefox itself, not in TSM.  Two files had items with openerTabId=self.id, and one had an openerTabId which didn't exist.  I discovered this while writing a quick script to convert the json files to a human-friendly text outline.)

FWIW, I'm on debian/stable, using firefox-esr 128.6.0esr-1~deb12u1, with TSM 7.0.1 from the Firefox extension store.

**Comment by ToyKeeper** [comment]
If it helps, I wrote a workaround and uploaded it:

https://github.com/ToyKeeper/Tab-Session-Manager/tree/tk-janky-upload-hack/tk-server

It's a quick dirty hack, and it requires the user to run a little web server on localhost, to receive and save session exports... but at least it works.

To be more precise, it works *for me* ... in my limited testing on a single version of Firefox in Linux.

I also included a tool to convert the json exports to human-friendly plain text outlines.  For example...

```
myuser @ myhost> json2txt.py mysession.json.gz
Session (4 windows, 20 / 129 tabs): Auto Save - Regularly
    - Date Exported: 2025-02-17 02:49:12 -0000
    - Date Started:  2025-02-17 01:02:03 -0000

Window 1 (10 / 42 tabs) [1826x1064+0+0]: Active Tab — Mozilla Firefox
    - Open Tab :: https://example.com/page1/
    * Closed (i.e. unloaded) Tab :: https://example.com/page2/
    + Closed Tab With Open Children
        - Open Tab :: https://example.com/page3/
        * Closed Tab :: https://example.com/page4/
        ! Active Tab :: https://example.com/page5/
    * TST Tab Group
        * Tab Groups don't need a URL
    ... (other items omitted) ...

Window 2 (4 / 12 tabs) [1822x1064+0+0]: Window Title — Mozilla Firefox
    ... (items omitted) ...
```

The tabs don't always appear in the same order that I see in the sidebar for Tree Style Tab, but that appears to be a bug in Tab Session Manager, and I didn't attempt to fix it.  The conversion script preserves the order and tree structure used in the json files, but the json files aren't always correct.

Anyway, the web server prints status as it receives requests.  Here's how it looks when it gets and saves a new session:

```
127.0.0.1 - - [17/Feb/2025 02:49:12] "OPTIONS /api/v1/session/post HTTP/1.1" 200 -
2025-02-17 02:49:12.890 localhost	==> /home/myuser/.tsm/2025/02/17/localhost.regular-myhost.2025-02-17_02:49:12.json.zst (39650 bytes)
127.0.0.1 - - [17/Feb/2025 02:49:12] "POST /api/v1/session/post HTTP/1.1" 200 -
```

By default, it compresses files with zstd.  If zstd isn't available, it uses gzip.

I didn't attempt to add any user-configurable options in TSM itself, to enable or disable this functionality or let the user select a URL to log to.  For now, it's just hardcoded to localhost port 4820, and if you want to change it you'll need to edit the source.

**Comment by sienori** [comment]
Thank you for reporting this issue. I am currently unable to reproduce the auto-save problem you're experiencing.

To help investigate this issue, could you please:
1. Reinstall Tab Session Manager
2. If the problem persists, please provide debug logs which can help identify the root cause

To enable debug logs: https://github.com/sienori/Tab-Session-Manager/wiki/Q&A-How-to-use-debug-mode

This information would be very helpful in resolving the issue. Thank you for your cooperation.

**Comment by ToyKeeper** [comment]
> To help investigate this issue, could you please:
> 1. Reinstall Tab Session Manager
> 2. If the problem persists...

I don't know about anyone else, but in my case, as I described earlier, I encountered this bug as a new user with a fresh install of Firefox and a fresh install of the latest version of TSM.

If I have time, I might try to do another fresh install with debug logs to see if there's anything useful in the log, but it's not a priority.  I've stopped using TSM and switched back to Ungoogled Chromium, because Firefox couldn't handle even a small fraction of the amount of tabs I use, and TSM was saving tabs in the wrong order in its session exports.

I got TSM to send the data using the patch I linked earlier, and a quick glance at that patch should reveal a lot about where the issue is, but the data itself wasn't intact.  TSM appears to use dicts instead of lists to store the tab tree, and although javascript dicts *can* preserve the key order, apparently TSM does some things which mess up the order somehow, so I'd almost always get mixed-up results.
